js-packages/@prestojs/util/src/identifiable.ts

```diff
@@ -0,0 +1,124 @@
+type SingleId = string | number;
+type CompoundId = { [fieldName: string]: SingleId };
+export type Id = SingleId | CompoundId;
+
+/**
+ * Interface for types that we can automatically extract a unique
+ * identifier from.
+ *
+ * If a `_pk` property exists (eg. from [ViewModelFactory](doc:ViewModelFactory))
+ * that will be used otherwise an `id` property will be used.
+ */
+export type Identifiable =
+    | {
+          id: Id;
+          _pk?: Id;
+      }
+    | {
+          id?: Id;
+          _pk: Id;
+      };
+
+/**
+ * Check if a value conforms to Identifiable
+ */
+export function isIdentifiable(item: any): item is Identifiable {
```

Comment:
honestly not entirely sure but I don't think so, see https://www.typescriptlang.org/play/?#code/C4TwDgpgBAyglgOwOYBsIEkAmUC8UDOwAToklAD5QICuAtgEYREDcAsAFCiRQDCA9rTB9qCTFlxQA3lADaAMzgQUmAHIBDWhABcBYqQC6O+MjTiAvm3YQAHkKLAoXaOLzHUGbJX6DhorJY4AegAqYI4oYKh0BGAmOTUAY2g5PiJHcAh8RwALNQcAd2gEtQQoNWpgATy4YpQUECgbYkSHNSgROABHaghwyLhMCBi4BSYoOSIBADo+vqi5MqgAAwB9MABrJagwSch7Bps4QiyACggkKfHJ2lkANUV8gFk+QZQAMRbUkH0TzD4ErT3CBPF5KD4JSpEEAASmhc2AuQKcDqUEY7XwEGwfARTHyR2gJWWAy2Oz4e1AUDxKLR1AxmBm7AigQ4NjsDicUUGwwUanoaFw4SgQsokkFQvF4oGOn8Yol4rW6wA-NLMJY5VAzLKRbL1QNlZy1er5RsVYaJRYOEFQnMeNkIAl1lARosAG5qFA9KAJPgIFJEWhZSqcobAEZwXloPrMqy2VIOOQiCFwH1O-BYENhiMQAA8ABUAHwnOCxWg6XPQnTFiA3I7B7nhvnQUWMyULE4AQirN3IlCcfAWXag7ZweAA5Hx6AAre3AUfQqQ68VECDAahEUrxFAYs0a2XL1frp0lqYDId4GgontH6tTBVnqjUOqWMxAA for what happens if make it generic

Reply:
Am not sure what I'm seeing here -- the output looks identical for both.

I realised that I missed the `: item is Identifiable` though so TS already has all the information it needs to do subsequent checks -- this is what I was thinking would allow this function to be potentially optimised away.

Reply:
Sorry I meant the errors you get if you change it to a generic